great_expectations/metrics/metric.py

```diff
@@ -0,0 +1,149 @@
+import re
+from typing import ClassVar, Final
+
+from typing_extensions import dataclass_transform
+
+from great_expectations.compatibility.pydantic import BaseModel, ModelMetaclass, root_validator
+from great_expectations.metrics.domain import AbstractClassInstantiationError, Domain
+from great_expectations.validator.metric_configuration import (
+    MetricConfiguration,
+    MetricConfigurationID,
+)
+
+ALLOWABLE_METRIC_MIXINS: Final[int] = 1
+
+
+class MixinTypeError(TypeError):
+    def __init__(self, class_name: str, mixin_superclass_name: str) -> None:
+        super().__init__(
+            f"`{class_name}` must use a single `{mixin_superclass_name}` subclass mixin."
+        )
+
+
+@dataclass_transform()
+class MetaMetric(ModelMetaclass):
+    def __new__(cls, name, bases, attrs):
+        # ensure a single Domain mixin is defined
+        if name != "Metric" and (
+            len(bases) != ALLOWABLE_METRIC_MIXINS + 1
+            or not any(issubclass(base_type, Domain) for base_type in bases)
+        ):
+            raise MixinTypeError(name, "Domain")
+        return super().__new__(cls, name, bases, attrs)
+
+
+class Metric(BaseModel, metaclass=MetaMetric):
+    """The abstract base class for defining all metrics.
+
+    A Metric represents a measurable property that can be computed over a specific domain
+    of data (e.g., a column, table, or column pair). All concrete metric implementations
+    must inherit from this class and specify their domain type as a mixin.
+
+    Examples:
+        A metric for column nullity values computed on each row:
+
+        >>> class Null(Metric, ColumnValues):
+        ...     ...
+
+        A metric for a single table row count value:
+
+        >>> class RowCount(Metric, Table):
+        ...     ...
+
+    Notes:
+        - The Metric class cannot be instantiated directly - it must be subclassed.
+        - Subclasses must specify a Domain type as a mixin.
+        - The subclass name and specified Domain type must be registered in the METRIC_REGISTRY.
```

Comment:
I think this no longer applies since you've removed the registry?

Reply:
Nice catch!